[HUDI-6369] Fix spacial curve with sample strategy fails when 0 or 1 rows only is incoming

parisni opened the pull request:
### Change Logs

Fixes #8934

see https://issues.apache.org/jira/browse/HUDI-6369

### Impact

Fix corner case for z-order clustering

### Risk level (write none, low medium or high below)
low

### Documentation Update
N/A

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Change Logs and Impact were stated clearly
- [x] Adequate tests were added if applicable
- [ ] CI passed

boneanxs commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/hudi/execution/RangeSample.scala:
```diff
@@ -258,7 +258,7 @@ object RangeSampleSort {
     val checkCols = orderByCols.filter(col => columnsMap(col) != null)
 
-    if (orderByCols.isEmpty || checkCols.isEmpty) {
+    if (orderByCols.isEmpty || checkCols.isEmpty || df.take(2).length < 2L) {
```
What abt to check rangeBounds' size? This way we don't need to read dataframe to get result again.

parisni replied:
yes makes sense. Also added a test case

boneanxs commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/hudi/execution/RangeSample.scala:
```diff
@@ -316,6 +316,8 @@ object RangeSampleSort {
         HoodieClusteringConfig.LAYOUT_OPTIMIZE_BUILD_CURVE_SAMPLE_SIZE.defaultValue.toString).toInt
       val sample = new RangeSample(zOrderBounds, sampleRdd)
       val rangeBounds = sample.getRangeBounds()
+      if (rangeBounds.size <= 1)
```
Can also fix `sortDataFrameBySampleSupportAllTypes`, it should have the same issue

parisni replied:
likely this one was not failing, still added a short path to avoid sorting

boneanxs replied:
if sort columns contain complex type, `sortDataFrameBySampleSupportAllTypes` will be used

parisni replied:
yes, the test has `height` column which is complex (array). But it did't trigger an error, a simple columns did.

boneanxs reviewed (commented):
Could you pls look into the test failure? It appeals new added test has errors. Also, this test can moved to `TestLayoutOptimization`.

Currently, we lack tests that cover the `sortDataFrameBySampleSupportAllTypes` function. It would be highly beneficial if you could include it as well.

parisni commented:
> Currently, we lack tests that cover the sortDataFrameBySampleSupportAllTypes function. It would be highly beneficial if you could include it as well.

Agreed, fill free to submit a patch, I am in vacation for a week

parisni commented:
> Currently, we lack tests that cover the sortDataFrameBySampleSupportAllTypes function. It would be highly beneficial if you could include it as well

- removed datasource test
- introduced new test for both sort methods to cover case
- verified test fails w/o the fix 

@boneanxs

boneanxs commented on hudi-client/hudi-spark-client/src/test/java/org/apache/spark/sql/hudi/execution/RangeSampleSortTest.java:
```diff
@@ -0,0 +1,41 @@
+
+import scala.collection.JavaConversions;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Fix this checkstyle

boneanxs reviewed (approved):
Overall LGTM, @danny0405 could you please also review this? Thanks!

bvaradar commented:
Looks good to me. Resurrecting this PR with rebase and minor test class rename.
